src/AspNet.Security.OAuth.Apple/Internal/DefaultAppleClientSecretGenerator.cs

```diff
@@ -22,39 +23,49 @@ internal sealed class DefaultAppleClientSecretGenerator : AppleClientSecretGener
         private readonly ILogger _logger;
         private readonly AppleKeyStore _keyStore;
         private readonly JwtSecurityTokenHandler _tokenHandler;
-
-        private string? _clientSecret;
-        private DateTimeOffset _expiresAt;
+        private readonly IMemoryCache _cache;
 
         public DefaultAppleClientSecretGenerator(
             [NotNull] AppleKeyStore keyStore,
             [NotNull] ISystemClock clock,
             [NotNull] JwtSecurityTokenHandler tokenHandler,
+            [NotNull] AppleMemoryCache cache,
```

Comment:
Why not just inject `IMemoryCache` directly, rather than create a whole separate cache?

Reply:
Directly inject `IMemoryCache` would make key collision and setting collision.
I followed this document: https://docs.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-3.1#use-setsize-size-and-sizelimit-to-limit-cache-size

Reply:
You could avoid a key collision by making the key namespace larger and more specific to the implementation (e.g. `DefaultAppleClientSecretGenerator+...`).

The limit should really be up to the owner of the integrating app via their existing cache setup, rather than creating an extra private cache that needs extra setup. If someone wants a size limit and it gets breached and throws that would be what they'd configured it to do.

Reply:
Okay. I'll change that.

Reply:
IMHO, things are starting to look quite messy here. Shouldn't we remove all these DI-resolved dependencies and move them to the options class? I suspect `JwtSecurityTokenHandler` is also one of those things people generating their own JWT tokens will register themselves, which may collide with the instance we use (e.g they could customize the claims mapping).

Reply:
The DI registrations all use `TryAdd...()` so that users' own things aren't overridden if they're registered already so we honour their own registrations if they're different:

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/8e4c19008f518f3730bab90a980e01347ba6f3d3/src/AspNet.Security.OAuth.Apple/AppleAuthenticationExtensions.cs#L81

Reply:
True, it means the Apple handler will use the `JwtSecurityTokenHandler` configured by the users - which may be configured to use things like claims mapping or have some of its methods overridden - and not the "default" one we'd register by default. My point is that users may affect the behavior of the Apple handler without even realizing it, which is not good.

Reply:
Is that something we can handle via docs as part of #391? I'd rather not radically (as in add lots of new options members that don't come from DI anymore) change the provider right before publishing the 3.1 version.

Reply:
Addressed as part of #484.